### ISSUE TITLE
tab-complete can mutate objects!

### ISSUE THREAD

**Issue by rsokl** [ipython/ipython#10720]
Consider the class.

```python
class X:
    T_cnt = []

    @property
    def T(self):
        self.T_cnt.append(len(self.T_cnt))
```

Thus `X.T_cnt` keeps track of the number of times that `X().T` was called.

The following demonstrates the bug:
```python
>>> x = X()
>>> x.<tab>
>>> x.<tab>
>>> X.T_cnt
[0, 1]
```

Each time I use <tab> to bring up the list of attributes of `x`, `x.T` is called under the hood. This occurs in both the ipython console and Jupyter notebook.

I am using iPython 6.1.0

**Comment by Carreau** [comment]
Probably related to https://github.com/davidhalter/jedi/issues/925 and https://github.com/ipython/ipython/issues/10580, I though on `@property` it was fixed, but that might be only on Jedi master.

**Comment by LLrsokl** [comment]
Actually, it appears that it only happens when `T` is a property, fortunately.

This can be pretty nasty for machine learning libraries, where `x.T` can produce a new tensor-node in a computational graph. I was getting super confused, when I saw all of these transpose operations in my graph.

**Comment by takluyver** [comment]
It's hard to do any useful introspection if we *guarantee* that tab completion has no side effects, because Python allows objects to override so much of their behaviour: almost anything we do might have side effects.

So we've generally tried to ensure that tab completion doesn't do things which are *likely* to have side effects (like calling regular methods), but allowed it to do things which usually don't (like attribute access).

I generally consider it bad design for property getters to have non-trivial side effects, because it's surprising both for introspection software and developers when that happens on attribute access.

**Comment by LLrsokl** [comment]
I agree that property getters should typically not have significant side effects - I certainly would never want to write code as what I contrived above.

That being said, this may be a recurring issue for libraries that create computational graphs. At the end of the day, this doesn't impact me much, especially now that I know what was causing it. I just wanted to make sure that you all knew about it.

**Comment by takluyver** [comment]
Thanks :-) It does sound like we (well, @Carreau) might be able to improve on this. Or might even already have done so.

**Comment by Carreau** [comment]
I'll have a look, but not soon. With PyBay comming and JupyterCon just
after it is tough.

It's on my todo list though.

**Comment by jcrist** [comment]
I just ran into this as well. I'm actually surprised that tab completion for attributes is more than just calling `dir` on the object. Dropping back to jedi `0.9.0` fixes the issue (may be useful when debugging).

```
(dask) jcrist dask $ cat foo.py
class Foo(object):
    @property
    def f(self):
        print("WOAH")

f = Foo()
(dask) jcrist dask $ ipython -i foo.py
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.0.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: f.<tab>  # single tab calls the property twice
WOAH
WOAH

In [2]: import jedi;jedi.__version__
Out[2]: '0.10.2'
```

Repeating same with jedi 0.9.0:

```
(dask) jcrist dask $ ipython -i foo.py
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.0.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: f.<tab>  # single tab calls nothing, just shows `f`

In [2]: import jedi;jedi.__version__
Out[2]: '0.9.0'
```

**Comment by mdeff** [comment]
Looks like this has been fixed in jedi 0.11 (davidhalter/jedi#925). At least upgrading jedi fixed the issue for me.
